### DISCUSSION TITLE
Does di-wise support constructor injection?

### DISCUSSION THREAD

**Discussion by shrink** [exuanbo/di-wise#12] [General]
Hi! I found this library via your comment on tysringe. Constructor injection is the most common form of dependency injection but from what I can tell, this library does not support it. Can you confirm whether or not this library supports constructor injection, and if not, is that a design decision or a feature you plan to implement in future? I think, for people coming from tsyringe, it would be very helpful to clarify the compatibility. Thank you :)

**Comment by exuanbo** [comment]
## without `Inject()` decorator

```ts
@Injectable()
class SomeService {
  constructor(dep: Dep) {}
}
```

This pattern is not supported because it requires TypeScript-only feature flag `emitDecoratorMetadata` together with legacy [`reflect-metadata`](https://www.npmjs.com/package/reflect-metadata) to collect the runtime information of the parameters. 

## with `Inject()` decorator

```ts
class SomeService {
  constructor(@Inject(Dep) dep: Dep) {}
}
```

This pattern is not supported for now, because the latest Stage 3 proposal [tc39/proposal-decorators](https://github.com/tc39/proposal-decorators) (introduced in TS 5) does not support decorating parameters [**yet**](https://github.com/tc39/proposal-decorators?tab=readme-ov-file#could-we-support-decorating-objects-parameters-blocks-functions-etc):

> Once we have validated this core approach, the authors of this proposal plan to come back and make proposals for more kinds of decorators. In particular, given the popularity of TypeScript parameter decorators, we are considering including parameter decorators in this proposal's initial version.

If the proposal changes in the future, I will follow up to implement the missing features.

---

So it is a design decision to follow the standardization and not to introduce legacy dependencies.

But you can always use the `inject()` function in constructors:

```ts
class SomeService {
  constructor(dep = inject(SomeToken)) {
    // inject() can be used anywhere during construction
  }
}
```

**Reply by exuanbo** [reply]
Thank you for asking this question :), I think my answer here should be added to the README for clarification.

**Reply by shrink** [reply]
Thank you for the quick answer, I appreciate it :) 

The idea of sticking with features that are available in Javascript, rather than through packages, is a great idea. Unfortunately, I think that means, at least for now, this package is not offering dependency injection: the `inject` function is not actually dependency injection, it is a service locator. There's a great Martin Fowler piece which compares a service locator to dependency injection and provides pros vs. cons for each ("[Inversion of Control Containers and the Dependency Injection pattern](https://martinfowler.com/articles/injection.html)").

For now, I'll stick with tsyringe, but keep an eye on this project, hopefully parameter decorators make it into Javascript someday soon!

**Reply by lppedd** [reply]
Hey folks! It's a been a few days I've been evaluating which library to use for DI. I come from Angular, where using `inject` is slowly becoming the way to go, so di-wise looks like a good candidate. 

@shrink I've looked at tsyringe and honestly it appears to be almost unmaintained, as issues are just sitting there indefinitely. Have you evaluated other libraries to migrate to?

@exuanbo just wondering, do you consider the library to be feature-complete? What are your maintenance plans going forward?

**Reply by shrink** [reply]
@lppedd yes, I've looked at various libraries over the years and I have always ended up staying with tsyringe. Although abandoned, it works without issue and no other library has come along with a better implementation. If a library comes along that is an improvement on tsyringe (I'd love to see first-class support for the [decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern)) then I would switch.

(The biggest problem with dependency inversion (or more broadly the SOLID principles) in the node/typescript/javascript world is that most dependencies aren't designed with it in mind. You're always going to have a codebase that is at best _sort of_ observing the principles but not really.)

However, as I write this comment, I'm checking out [Inversify](https://inversify.io) again. Historically tsyringe and Inversify were the most popular implementations. Last time I looked, both tsyringe and Inversify were abandoned and I had passed on Inversify for some (long forgotten) reason. Inversify is now under active development with a new major version. Reviewing the Inversify documentation today it looks like a promising alternative to tsyringe.

I'm going to try out Inversify again and that's where I'd recommend you start. I think this library (di-wise) is very well written but the lack of constructor injection makes it a non-starter for me.

**Reply by lppedd** [reply]
@shrink thank you! di-wise code is actually excellent, as you wrote. No scope creep, minimalistic approach, use of weak refs. A point in its favor is whatever happens to it forking is a no-brainer. On the other hand, using standard decorators is a long term set back, but I understand the _why_.

Regarding InversifyJS, I doubt I'll be able to use it since it requires decorator metadata and I'm using ESBuild. The same applies to TSyringe tho, but at least it's easier to patch.

> If a library comes along that is an improvement on tsyringe

I think you might be interested into https://redi.wzhu.dev/en-US, which is built by people that previously worked on [NG-ZORRO](https://github.com/NG-ZORRO/ng-zorro-antd), a UI library for Angular. It supports constructor injection, and it _does not_ use metadata. I'm evaluating it alongside di-wise.

**Reply by lppedd** [reply]
While @exuanbo is away, I've been working on additional refactorings/features, to better integrate DI into our VS Code extensions. I'm committing to https://github.com/lppedd/di-wise-neo/commits/feat/neo/, but my target is to port back (almost) everything to di-wise.

I'm working on a new `ExistingProvider<T>`, and I will then switch back (in a separate branch) to TypeScript's experimental decorators (as other deps we're using - unfortunately - use them) with the goal of adding constructor injection via parameter decorators.

**Reply by lppedd** [reply]
I've reverted to using TS experimental decorators, and implemented constructor and method injection.
I've removed support for property injection as with the experimental decorators they're dangerous to use (their initialization is done after construction, instead of during construction as with ECMAScript decorators).

@shrink see https://github.com/lppedd/di-wise-neo/blob/4e7112957b2d5bf5b0bf7670eb66f5d62c94fe50/src/__tests__/container.spec.ts#L186-L327 for some examples.
